### PR TITLE
[Feature] k-hop Subgraph Extraction

### PR DESCRIPTION
**Pull request by mufeili**
## Description
Subgraph induced on nodes in the k-hop neighborhood

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

**Review by BarclayII — APPROVED** [review]
Looks good.  One question though: can I conveniently know which nodes in the subgraph are the given nodes?

File: python/dgl/subgraph.py [hunk 1]
```diff
@@ -552,6 +552,254 @@ def out_subgraph(graph, nodes, *, relabel_nodes=False, store_ids=True):
...
+    We can expand a set of nodes by including the predecessors of them. From a
+    specified node, a k-hop in subgraph is obtained by first repeating the node set
+    expansion for k times and then creating a node induced subgraph. In addition to
+    extracting the subgraph, DGL also copies the features of the extracted nodes and
+    edges to the resulting graph. The copy is *lazy* and incurs data movement only
+    when needed.
+
+    If the graph is heterogeneous, DGL extracts a subgraph per relation and composes
+    them as the resulting graph. Thus the resulting graph has the same set of relations
+    as the input one.
```
[the remark this answers is not in the document]
Comment (mufeili) [reply python/dgl/subgraph.py:568]: There will be an empty graph for that relation. This is the behavior of `node_graph`. See the example below.

```python
import dgl

g = dgl.heterograph({('user', 'plays', 'game'): ([0, 1, 1, 2], [0, 0, 2, 1]), ('user', 'follows', 'user'): ([0, 1, 1], [1, 2, 2])})
dgl.node_subgraph(g, {'user': [1, 2]})
# Graph(num_nodes={'game': 0, 'user': 2},
#             num_edges={('user', 'follows', 'user'): 2, ('user', 'plays', 'game'): 0},
#             metagraph=[('user', 'user', 'follows'), ('user', 'game', 'plays')])
```

File: tests/compute/test_subgraph.py [hunk 2]
```diff
@@ -460,3 +460,83 @@ def test_subgraph_message_passing():
     g.ndata['x'] = F.copy_to(F.randn((5, 6)), F.cpu())
     sg = g.subgraph([1, 2, 3]).to(F.ctx())
     sg.update_all(lambda edges: {'x': edges.src['x']}, lambda nodes: {'y': F.sum(nodes.mailbox['x'], 1)})
+
+@parametrize_dtype
+def test_khop_in_subgraph(idtype):
```
[the remark this answers is not in the document]
Comment (mufeili) [reply tests/compute/test_subgraph.py:465]: Done

File: tests/compute/test_subgraph.py [hunk 3]
```diff
@@ -460,3 +460,83 @@ def test_subgraph_message_passing():
...
+    assert len(sg.etypes) == 2
+    u, v = sg['follows'].edges()
+    edge_set = set(zip(list(F.asnumpy(u)), list(F.asnumpy(v))))
+    assert edge_set == {(0, 0), (1, 0)}
+    u, v = sg['plays'].edges()
+    edge_set = set(zip(list(F.asnumpy(u)), list(F.asnumpy(v))))
+    assert edge_set == {(0, 1)}
+
+@parametrize_dtype
+def test_khop_out_subgraph(idtype):
```
[the remark this answers is not in the document]
Comment (mufeili) [reply tests/compute/test_subgraph.py:519]: Done

**Comment by mufeili** [comment]
@BarclayII 
1. I've extended `node` to a node set, i.e. multiple nodes as in the case of `node_subgraph`.
2. For `F.unique`, I think torch allows returning the inverse map, but I'm not sure if this also holds for TensorFlow and MXNet. From the [implementation of VertexSubgraph](https://github.com/dmlc/dgl/blob/5a2451047fd8156a86bcfa6fd157a3f31328726e/src/graph/unit_graph.cc#L372), it seems that the relative order of the input node set is preserved for each node type.

**Comment by mufeili** [comment]
@BarclayII I've merged the PR first and we can update it if there are any additional requirements.

**Pull request merged**